File: php/tests/EffectivePropertySetTest.php
<?php

declare(strict_types=1);

namespace Oojs\Tests;

use PHPUnit\Framework\TestCase;

/**
 * §9.5 — Effective property set (effectiveProperties / effectiveRequired).
 */
class EffectivePropertySetTest extends TestCase
{
    use TestHelpers;

    public function test_inherits_parent_properties(): void
    {
        [$r, $schema] = self::animalRegistry();
        $eff = $schema->types['Dog']->effectiveProperties();
        $this->assertArrayHasKey('name', $eff);   // from Animal
        $this->assertArrayHasKey('age', $eff);    // from Animal
        $this->assertArrayHasKey('breed', $eff);  // own
    }

    public function test_effective_required_union(): void
    {
        [$r, $schema] = self::animalRegistry();
        $req = $schema->types['Dog']->effectiveRequired();
        $this->assertContains('name', $req);   // inherited
        $this->assertContains('breed', $req);  // own
    }
}

File: java/src/test/java/org/oojs/EffectivePropertySetTest.java
package org.oojs;

import org.junit.jupiter.api.Test;
import org.oojs.model.Schema;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class EffectivePropertySetTest extends TestHelpers {

    @Test void inherits_parent_properties() {
        Schema schema = (Schema) animalRegistry()[1];
        var eff = schema.types.get("Dog").effectiveProperties();
        assertTrue(eff.containsKey("name"));
        assertTrue(eff.containsKey("age"));
        assertTrue(eff.containsKey("breed"));
    }

    @Test void effective_required_union() {
        Schema schema = (Schema) animalRegistry()[1];
        List<String> req = schema.types.get("Dog").effectiveRequired();
        assertTrue(req.contains("name"));
        assertTrue(req.contains("breed"));
    }
}
